ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java

```diff
@@ -4521,7 +4523,9 @@ public static void handleDirectInsertTableFinalPath(Path specPath, String unionS
           for (int i = 0; i < partitionCount; ++i) {
             String nextPart = mdis.readUTF();
             Utilities.FILE_OP_LOGGER.debug("Looking at dynamic partition {}", nextPart);
-            dynamicPartitionSpecs.add(nextPart);
+            if (!dynamicPartitionSpecs.containsKey(nextPart)) {
+              dynamicPartitionSpecs.put(nextPart, new ArrayList<>());
```

Comment:
nit: why List?

Reply:
At the end it is used as a List in PartitionDetails, I could use a set here, but I would convert it anyway to a list.

Reply:
What is the expected order of the items in the list?
Do we really need a list there?

Reply:
Not really, but it starts a refactor domino. I started it but reverted, I do not want to pollute this PR.